soc: nxp: imxrt: exclude mpu_regions.c when ARM_MPU disabled

GrixaYrev opened the pull request:
Option NXP_IMX_RT_USER_MPU_REGIONS exclude soc file mpu_regions.c and allow user define custom mpu_regions in custom file. 

Fixes #70920

For test this PR i use:

- file prj.conf:
```
...
CONFIG_FLASH_SIZE=5120 
CONFIG_FLASH_BASE_ADDRESS=0x80100000
CONFIG_SRAM_SIZE=26624
CONFIG_SRAM_BASE_ADDRESS=0x80600000
...
```

- file with custom MPU regions:
```
#include <zephyr/devicetree.h>
#include <zephyr/arch/arm/mpu/arm_mpu.h>

#define REGION_RAM_EXEC_ATTR(size) \
{ \
	(NORMAL_OUTER_INNER_WRITE_BACK_WRITE_READ_ALLOCATE_NON_SHAREABLE | \
	 size | P_RW_U_NA_Msk) \
}

static const struct arm_mpu_region mpu_regions[] = {
	/* Region 1 */
	MPU_REGION_ENTRY("SRAM_0",
			 CONFIG_SRAM_BASE_ADDRESS,
			 REGION_RAM_EXEC_ATTR(REGION_32M)),
};

const struct arm_mpu_config mpu_config = {
	.num_regions = ARRAY_SIZE(mpu_regions),
	.mpu_regions = mpu_regions,
};
```

danieldegrasse reviewed (commented):
While I understand the goal here, I'm not sure I agree with the approach. What prevents you from defining the MPU regions you need using the `zephyr,memory-region` compatible and `zephyr,memory-attr` property?

Is there a specific MPU region the IMX RT MPU regions file defines that you need removed?

GrixaYrev commented:
@danieldegrasse, main trouble is checking of `CONFIG_SRAM_SIZE` in file **zephyr\include\zephyr\arch\arm\cortex_m\arm_mpu_mem_cfg.h** (see issue #70920).
Also i think user should can setup mpu regions flexible, because their number is limited. We can not provide all possible memory configurations, so standard file should configure most common case, but user should can use custom file for special configuration.

danieldegrasse commented:
> @danieldegrasse, main trouble is checking of `CONFIG_SRAM_SIZE` in file **zephyr\include\zephyr\arch\arm\cortex_m\arm_mpu_mem_cfg.h** (see issue #70920). Also i think user should can setup mpu regions flexible, because their number is limited. We can not provide all possible memory configurations, so standard file should configure most common case, but user should can use custom file for special configuration.

Ok- thank you for clarifying. Looking at that issue, I guess you would like to use an SRAM size that is not a multiple supported by the ARM MPU? Perhaps we could solve this by rounding the MPU region size up to the nearest supported size? Or would this not work for your use case?

GrixaYrev commented:
@danieldegrasse, rounding could solve main trouble probably, but MPU will not work correctly in this case, because start address should be rounded to region size, if i am not wrong. Now i use no MPU regions (all SDRAM configured as executable, readable and writeable RAM), but i want to use it in future. So i want to find flexible solution for my issue.

danieldegrasse commented:
> @danieldegrasse, rounding could solve main trouble probably, but MPU will not work correctly in this case, because start address should be rounded to region size, if i am not wrong. Now i use no MPU regions (all SDRAM configured as executable, readable and writeable RAM), but i want to use it in future. So i want to find flexible solution for my issue.

So I think what I am unclear about is how removing the MPU definition file would solve your issue here. Either you don't use the MPU, or you enable the MPU and now you have some restrictions on the size/alignment of your SRAM region. Maybe it would help if you could provide an outline of a case where *replacing* the MPU file would allow you to accomplish something you currently cannot?

danieldegrasse commented:
> @danieldegrasse, I want to set follow memory layout in **prj.conf**:
> 
> ```
> CONFIG_FLASH_SIZE=5120 
> CONFIG_FLASH_BASE_ADDRESS=0x80100000
> CONFIG_SRAM_SIZE=26624
> CONFIG_SRAM_BASE_ADDRESS=0x80600000
> ```
> 
> I have error "_Unsupported sram size configuration_" when i set CONFIG_SRAM_SIZE more than 16M and less than 32M.
> 
> I have this error, even i switch off MPU:
> 
> ```
> CONFIG_ARM_MPU=n
> CONFIG_MPU=n
> ```
> 
> This error is described in `zephyr\include\zephyr\arch\arm\cortex_m\arm_mpu_mem_cfg.h:72`.
> 
> It is not MPU trouble, it is code trouble,

Ok, thank you for this example, the issue is easier to follow now. So it seems like we just need to stop compiling the MPU regions file if `CONFIG_ARM_MPU=n`. Locally this change works for me:

```diff
diff --git a/soc/nxp/imxrt/CMakeLists.txt b/soc/nxp/imxrt/CMakeLists.txt
index ecdb894bfe3..09e90e87fa5 100644
--- a/soc/nxp/imxrt/CMakeLists.txt
+++ b/soc/nxp/imxrt/CMakeLists.txt
@@ -16,7 +16,7 @@ if(CONFIG_SOC_SERIES_IMXRT10XX OR CONFIG_SOC_SERIES_IMXRT11XX)
   if(CONFIG_EXTERNAL_MEM_CONFIG_DATA)
     set(boot_hdr_xmcd_data_section ".boot_hdr.xmcd_data")
   endif()
-  zephyr_sources(mpu_regions.c)
+  zephyr_sources_ifdef(CONFIG_ARM_MPU mpu_regions.c)
   zephyr_linker_section_configure(
     SECTION .rom_start
     INPUT ".boot_hdr.conf"
```

GrixaYrev commented:
I think i should add this change to commit, but it is restricted solution. I want to use MPU in future, and i will have this error again. So I suggest to join your change and my change.

danieldegrasse commented:
> I think i should add this change to commit, but it is restricted solution. I want to use MPU in future, and i will have this error again. So I suggest to join your change and my change.

I'm still unclear about the use case for the MPU on your side. Do you plan to only protect part of the region you use as SRAM? The reason I am hesitant to add a Kconfig to exclude the MPU file is because this MPU file effectively mirrors the one used for most ARM cores, plus an additional region. If you want to use an MPU in Zephyr on an ARMv7 platform, you will need SRAM regions aligned to the appropriate size boundaries. By adding a Kconfig to specifically exclude this for the iMX.RT series, it becomes possible to define MPU regions that don't fit the alignment requirements Zephyr MPU support is designed around on ARMv7. Is there a part of the Zephyr MPU support you are looking to use? If not, it truthfully seems to me like configuring the MPU directly and bypassing Zephyr is more logical.

GrixaYrev commented:
I plan use usermode. But i don't have enough experience in it. May be you are right and it is unavailable with this memory layout.
OK, i change this pr for your solution and will return for my variant when start to use usermode.

dleach02 commented:
@GrixaYrev What do you want to do with this PR?

GrixaYrev commented:
Sorry for pause, i want to return to this issue in November. I think this PR should be closed and i will create new PR. Is it right? Or should i change this PR?

danieldegrasse commented:
> Sorry for pause, i want to return to this issue in November. I think this PR should be closed and i will create new PR. Is it right? Or should i change this PR?

You can update this PR when you return to it, no rush.

GrixaYrev committed:
soc: nxp: imxrt: exclude mpu_regions.c when ARM_MPU disabled

When option ARM_MPU is disabled exclude soc\nxp\imxrt\mpu_regions.c.
It is needed to remove constraints of SRAM and FLASH size.
Fixes #70920

Signed-off-by: Grixa Yrev <GrixaYrev@yandex.ru>

danieldegrasse reviewed (approved):
Shouldn't break anything, although I'm unclear what this fixes (other than marginally reducing build time when `CONFIG_ARM_MPU` is disabled?)

No problem with the PR, just trying to understand the use case

GrixaYrev commented:
It remove constraints of SRAM and FLASH size, which defined in `zephyr\include\zephyr\arch\arm\cortex_m\arm_mpu_mem_cfg.h` which included to `zephyr\soc\nxp\imxrt\mpu_regions.c`.

decsny commented:
ARM_MPU should not be being disabled, it indicates the mcu has an arm mpu, its not meant to be application configured.

danieldegrasse commented:
> ARM_MPU should not be being disabled, it indicates the mcu has an arm mpu, its not meant to be application configured.

How else would a user disable support for the MPU within their application code?

decsny commented:
> > ARM_MPU should not be being disabled, it indicates the mcu has an arm mpu, its not meant to be application configured.
> 
> How else would a user disable support for the MPU within their application code?

What does this mean? Disable support for the MPU in their application code ? which part of their code would do this and what does it mean to "disable support"

danieldegrasse commented:
> What does this mean? Disable support for the MPU in their application code ? which part of their code would do this and what does it mean to "disable support"

Disable support for the MPU within Zephyr, so that Zephyr won't configure the MPU nor allow features that require it to be used. I could see a user doing this if they wanted to manage the MPU outside of Zephyr for some reason.

decsny commented:
> > What does this mean? Disable support for the MPU in their application code ? which part of their code would do this and what does it mean to "disable support"
> 
> Disable support for the MPU within Zephyr when building their application, so that Zephyr won't configure the MPU nor allow features that require it to be used. I could see a user doing this if they wanted to manage the MPU outside of Zephyr for some reason.

it still is not matching what that kconfig is supposed to mean, according to the description, so right now doing that is a hack with no guarantee to do what youre saying, but, the PR here doesn't really affect anything so im not blocking this PR, just pointing out that theres no guarantee to anyone doing that , that its a valid thing to do, because its not, and even if it somehow works now, it won't necessarily in the future. Because the contract of that Kconfig is not like "turn this on to enable the MPU", its "select this if the platform has an MPU"

GrixaYrev commented:
I think we should use CONFIG_MPU, but it is not available. May be we can change CONFIG_MPU and CONFIG_ARM_MPU dependence? May be should change `select MPU` in **ARM_MPU** to `default y if ARM_MPU` in **MPU**?

danieldegrasse commented:
> I think we should use CONFIG_MPU, but it is not available. May be we can change CONFIG_MPU and CONFIG_ARM_MPU dependence? May be should change `select MPU` in **ARM_MPU** to `default y if ARM_MPU` in **MPU**?

I agree `CONFIG_MPU` would make more sense here, but the way the dependency chain is set up you'd need to make a larger change to allow users to disable CONFIG_MPU outright. I think just this change is fine for now

dleach02 commented:
The kconfig stuff gets a bit twisted since you have a CPU_HAS_ARM_MPU which is the indicator that one exists. This implies that the CONFIG_ARM_MPU is "turn it on/off". The ARM_MPU usually gets set in the board definition but seems okay to let the application override it for developer boards.

Adding yet another CONFIG_MPU would complicate things more but would be the architecture agnostic way for dealing with MPU enabled/disabled.
